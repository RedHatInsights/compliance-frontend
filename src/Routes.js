import { Route, Switch, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import React from 'react';
import asyncComponent from './Utilities/asyncComponent';
import some from 'lodash/some';

/**
 * Aysnc imports of components
 *
 * https://webpack.js.org/guides/code-splitting/
 * https://reactjs.org/docs/code-splitting.html
 *
 * pros:
 *      1) code splitting
 *      2) can be used in server-side rendering
 * cons:
 *      1) nameing chunk names adds unnecessary docs to code,
 *         see the difference with DashboardMap and InventoryDeployments.
 *
 */
const CompliancePolicies = asyncComponent(() =>
    import(/* webpackChunkName: "CompliancePolicies" */ './SmartComponents/CompliancePolicies/CompliancePolicies')
);
const ComplianceSystems = asyncComponent(() =>
    import(/* webpackChunkName: "ComplianceSystems" */ './SmartComponents/ComplianceSystems/ComplianceSystems')
);
const PolicyDetails = asyncComponent(() =>
    import(/* webpackChunkName: "PolicyDetails" */ './SmartComponents/PolicyDetails/PolicyDetails')
);
const SystemDetails = asyncComponent(() =>
    import(/* webpackChunkName: "SystemDetails" */ './SmartComponents/SystemDetails/SystemDetails')
);

export const paths = {
    compliancePolicies: '/policies',
    complianceSystems: '/systems',
    complianceSystemsInventoryDetail: '/systems/:inventoryId',
    policyDetails: '/policies/:policy_id',
    policyDetailsInventoryDetail: '/policies/:policy_id/:inventoryId',
    systemDetails: '/systems/:inventoryId'
};

type Props = {
    childProps: any
};

const InsightsRoute = ({ component: Component, rootClass, ...rest }) => {
    const root = document.getElementById('root');
    root.removeAttribute('class');
    root.classList.add(`page__${rootClass}`, 'pf-c-page__main');
    root.classList.add(`page__${rootClass}`, 'pf-l-page__main');
    root.setAttribute('role', 'main');

    return (<Component { ...rest } />);
};

InsightsRoute.propTypes = {
    component: PropTypes.func,
    rootClass: PropTypes.string
};

/**
 * the Switch component changes routes depending on the path.
 *
 * Route properties:
 *      exact - path must match exactly,
 *      path - https://prod.foo.redhat.com:1337/insights/compliance/rules
 *      component - component to be rendered when a route has been chosen.
 */
export const Routes = (props: Props) => {
    const path = props.childProps.location.pathname;

    return (
        <Switch>
            <Route exact path={paths.compliancePolicies} component={CompliancePolicies} />
            <Route exact path={paths.complianceSystems} component={ComplianceSystems} />
            <Route path={paths.complianceSystemsInventoryDetail} component={SystemDetails} />
            <Route path={paths.policyDetailsInventoryDetail} component={SystemDetails} />
            <Route exact path={paths.policyDetails} component={PolicyDetails} />
            <Route path={paths.systemDetails} component={SystemDetails} />

            { /* Finally, catch all unmatched routes */ }
            <Route render={ () => some(paths, p => p === path) ? null : (<Redirect to={paths.compliancePolicies}/>) }/>
        </Switch>
    );
};
