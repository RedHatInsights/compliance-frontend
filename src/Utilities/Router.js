import React, { useEffect, Suspense, Fragment } from 'react';
import propTypes from 'prop-types';
import {
  Route as ReactRoute,
  MemoryRouter,
  Switch,
  Redirect,
  useHistory,
  useLocation,
  matchPath,
} from 'react-router-dom';
import useChrome from '@redhat-cloud-services/frontend-components/useChrome';
import { WithPermission } from 'PresentationalComponents';
import useDocumentTitle from 'Utilities/hooks/useDocumentTitle';
import { routes as defaultRoutes } from '../Routes';

const Route = (route) => {
  const {
    component: Component,
    modal,
    path,
    props = {},
    title,
    requiredPermissions,
  } = route;
  const location = useLocation();
  const setTitle = useDocumentTitle();
  const isCurrent = !!matchPath(location.pathname, { path, exact: true });
  const requiresTitleEntity = title.includes('$entityTitle');
  const routeProps = {
    exact: true,
    key: `${!modal ? 'fullpage' : 'modal'}-route-${path.replace('/', '-')}`,
    path,
  };
  const componentProps = {
    ...props,
    route: { ...route, isCurrent, setTitle },
  };

  useEffect(() => {
    isCurrent && !requiresTitleEntity && setTitle(title);
  });

  return (
    <ReactRoute {...routeProps}>
      {requiredPermissions ? (
        <WithPermission requiredPermissions={requiredPermissions}>
          <Component {...componentProps} />
        </WithPermission>
      ) : (
        <Component {...componentProps} />
      )}
    </ReactRoute>
  );
};

Route.propTypes = {
  component: propTypes.node,
  modal: propTypes.bool,
  path: propTypes.string,
  props: propTypes.object,
  title: propTypes.string,
  requiredPermissions: propTypes.array,
};

const Router = ({ routes }) => {
  const location = useLocation();
  const background = location?.state?.background;
  const path = location.pathname;
  const fullPageRoutes = routes.filter((route) => !route.modal);
  const modalRoutes = routes.filter((route) => route.modal);
  const paths = routes.map((route) => route.path);
  const defaultRedirectRender = () =>
    paths.some((p) => p === path) ? null : <Redirect to="/reports" />;

  return (
    <Suspense fallback={Fragment}>
      <Switch location={background || location}>
        {fullPageRoutes.map(Route)}
        <ReactRoute render={defaultRedirectRender} />
      </Switch>

      {modalRoutes.map(Route)}
    </Suspense>
  );
};

Router.propTypes = {
  routes: propTypes.array,
};

const RedirectComponent = ({ chromeHistory, isBeta, ...route }) => {
  useEffect(() => {
    const path = `/insights/compliance${route.path}`;

    // NOTE: This may stop working with the (chrome/platform) migration to react router v6.
    // Hence the fallback.
    if (chromeHistory) {
      chromeHistory.push(path);
    } else {
      location.pathname = path;
    }
  }, []);
  return <></>;
};

const InsightsRedirect = (route) => (
  <ReactRoute path={route.path}>
    <RedirectComponent {...route} />
  </ReactRoute>
);

RedirectComponent.propTypes = {
  chromeHistory: propTypes.object,
  isBeta: propTypes.string,
};

export const FederatedModuleRouter = ({ routes = defaultRoutes, children }) => {
  const { chromeHistory, isBeta } = useChrome();

  return (
    <MemoryRouter>
      <Suspense fallback={Fragment}>
        <Switch>
          {routes.map((route, idx) => (
            <InsightsRedirect
              key={'redirect-' + idx}
              {...{ ...route, chromeHistory, isBeta }}
            />
          ))}
        </Switch>
        {children}
      </Suspense>
    </MemoryRouter>
  );
};

FederatedModuleRouter.propTypes = {
  routes: propTypes.array,
  children: propTypes.node,
};

export const useLinkToBackground = (fallbackRoute) => {
  const location = useLocation();
  const history = useHistory();

  return (args) => {
    const background = location?.state?.background;

    history.push({
      pathname: background ? background.pathname : fallbackRoute,
      hash: background ? background.hash : undefined,
      ...args,
    });
  };
};

export const useAnchor = (defaultAnchor = '') => {
  const location = useLocation();
  const hash =
    location.hash && location.hash.length > 0 ? location.hash : undefined;
  return (hash || defaultAnchor).replace('#', '');
};

export default Router;
