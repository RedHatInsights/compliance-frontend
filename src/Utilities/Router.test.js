import Router from './Router';
import { routes } from '@/Routes';
import { useLocation } from 'react-router-dom';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn(),
}));

describe('Router', () => {
  it('expect to render without error', () => {
    useLocation.mockImplementation(() => ({
      pathname: '/reports',
    }));
    const wrapper = shallow(<Router {...{ routes }} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
