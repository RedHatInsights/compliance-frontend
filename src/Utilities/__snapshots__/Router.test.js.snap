// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router expect to render without error 1`] = `
<Suspense
  fallback={Symbol(react.fragment)}
  routes={
    Array [
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "path": "/scappolicies",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:policy:read",
        ],
        "title": "SCAP policies",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "modal": true,
        "path": "/scappolicies/new",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:policy:create",
        ],
        "title": "SCAP policies",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "defaultTitle": "SCAP policies",
        "path": "/scappolicies/:policy_id",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:policy:read",
        ],
        "title": "$entityTitle - SCAP policies",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "defaultTitle": "SCAP policies",
        "modal": true,
        "path": "/scappolicies/:policy_id/edit",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:policy:update",
        ],
        "title": "$entityTitle - SCAP policies",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "modal": true,
        "path": "/scappolicies/:policy_id/delete",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:policy:delete",
        ],
        "title": "Delete policy - SCAP policies",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "path": "/reports",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:report:read",
        ],
        "title": "Reports",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "defaultTitle": "Reports",
        "path": "/reports/:report_id",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:report:read",
        ],
        "title": "Report: $entityTitle - Reports",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "modal": true,
        "path": "/reports/:report_id/delete",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:report:delete",
        ],
        "title": "Delete report - Reports",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "defaultTitle": "Reports",
        "modal": true,
        "path": "/reports/:report_id/pdf",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:report_export:read",
        ],
        "title": "Export report - Reports",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "path": "/systems",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:system:read",
        ],
        "title": "Compliance systems",
      },
      Object {
        "component": Object {
          "$$typeof": Symbol(react.lazy),
          "_init": [Function],
          "_payload": Object {
            "_result": [Function],
            "_status": -1,
          },
        },
        "defaultTitle": "Compliance systems",
        "path": "/systems/:inventoryId",
        "requiredPermissions": Array [
          "compliance:*:*",
          "compliance:system:read",
        ],
        "title": "$entityTitle - Compliance systems",
      },
    ]
  }
>
  <Switch
    location={
      Object {
        "pathname": "/reports",
      }
    }
  >
    <Route
      exact={true}
      key="fullpage-route--scappolicies"
      path="/scappolicies"
    >
      <WithPermission
        requiredPermissions={
          Array [
            "compliance:*:*",
            "compliance:policy:read",
          ]
        }
      >
        <lazy
          route={
            Object {
              "component": Object {
                "$$typeof": Symbol(react.lazy),
                "_init": [Function],
                "_payload": Object {
                  "_result": [Function],
                  "_status": -1,
                },
              },
              "isCurrent": false,
              "path": "/scappolicies",
              "requiredPermissions": Array [
                "compliance:*:*",
                "compliance:policy:read",
              ],
              "setTitle": [Function],
              "title": "SCAP policies",
            }
          }
        />
      </WithPermission>
    </Route>
    <Route
      exact={true}
      key="fullpage-route--scappolicies/:policy_id"
      path="/scappolicies/:policy_id"
    >
      <WithPermission
        requiredPermissions={
          Array [
            "compliance:*:*",
            "compliance:policy:read",
          ]
        }
      >
        <lazy
          route={
            Object {
              "component": Object {
                "$$typeof": Symbol(react.lazy),
                "_init": [Function],
                "_payload": Object {
                  "_result": [Function],
                  "_status": -1,
                },
              },
              "defaultTitle": "SCAP policies",
              "isCurrent": false,
              "path": "/scappolicies/:policy_id",
              "requiredPermissions": Array [
                "compliance:*:*",
                "compliance:policy:read",
              ],
              "setTitle": [Function],
              "title": "$entityTitle - SCAP policies",
            }
          }
        />
      </WithPermission>
    </Route>
    <Route
      exact={true}
      key="fullpage-route--reports"
      path="/reports"
    >
      <WithPermission
        requiredPermissions={
          Array [
            "compliance:*:*",
            "compliance:report:read",
          ]
        }
      >
        <lazy
          route={
            Object {
              "component": Object {
                "$$typeof": Symbol(react.lazy),
                "_init": [Function],
                "_payload": Object {
                  "_result": [Function],
                  "_status": -1,
                },
              },
              "isCurrent": true,
              "path": "/reports",
              "requiredPermissions": Array [
                "compliance:*:*",
                "compliance:report:read",
              ],
              "setTitle": [Function],
              "title": "Reports",
            }
          }
        />
      </WithPermission>
    </Route>
    <Route
      exact={true}
      key="fullpage-route--reports/:report_id"
      path="/reports/:report_id"
    >
      <WithPermission
        requiredPermissions={
          Array [
            "compliance:*:*",
            "compliance:report:read",
          ]
        }
      >
        <lazy
          route={
            Object {
              "component": Object {
                "$$typeof": Symbol(react.lazy),
                "_init": [Function],
                "_payload": Object {
                  "_result": [Function],
                  "_status": -1,
                },
              },
              "defaultTitle": "Reports",
              "isCurrent": false,
              "path": "/reports/:report_id",
              "requiredPermissions": Array [
                "compliance:*:*",
                "compliance:report:read",
              ],
              "setTitle": [Function],
              "title": "Report: $entityTitle - Reports",
            }
          }
        />
      </WithPermission>
    </Route>
    <Route
      exact={true}
      key="fullpage-route--systems"
      path="/systems"
    >
      <WithPermission
        requiredPermissions={
          Array [
            "compliance:*:*",
            "compliance:system:read",
          ]
        }
      >
        <lazy
          route={
            Object {
              "component": Object {
                "$$typeof": Symbol(react.lazy),
                "_init": [Function],
                "_payload": Object {
                  "_result": [Function],
                  "_status": -1,
                },
              },
              "isCurrent": false,
              "path": "/systems",
              "requiredPermissions": Array [
                "compliance:*:*",
                "compliance:system:read",
              ],
              "setTitle": [Function],
              "title": "Compliance systems",
            }
          }
        />
      </WithPermission>
    </Route>
    <Route
      exact={true}
      key="fullpage-route--systems/:inventoryId"
      path="/systems/:inventoryId"
    >
      <WithPermission
        requiredPermissions={
          Array [
            "compliance:*:*",
            "compliance:system:read",
          ]
        }
      >
        <lazy
          route={
            Object {
              "component": Object {
                "$$typeof": Symbol(react.lazy),
                "_init": [Function],
                "_payload": Object {
                  "_result": [Function],
                  "_status": -1,
                },
              },
              "defaultTitle": "Compliance systems",
              "isCurrent": false,
              "path": "/systems/:inventoryId",
              "requiredPermissions": Array [
                "compliance:*:*",
                "compliance:system:read",
              ],
              "setTitle": [Function],
              "title": "$entityTitle - Compliance systems",
            }
          }
        />
      </WithPermission>
    </Route>
    <Route
      render={[Function]}
    />
  </Switch>
  <Route
    exact={true}
    key="modal-route--scappolicies/new"
    path="/scappolicies/new"
  >
    <WithPermission
      requiredPermissions={
        Array [
          "compliance:*:*",
          "compliance:policy:create",
        ]
      }
    >
      <lazy
        route={
          Object {
            "component": Object {
              "$$typeof": Symbol(react.lazy),
              "_init": [Function],
              "_payload": Object {
                "_result": [Function],
                "_status": -1,
              },
            },
            "isCurrent": false,
            "modal": true,
            "path": "/scappolicies/new",
            "requiredPermissions": Array [
              "compliance:*:*",
              "compliance:policy:create",
            ],
            "setTitle": [Function],
            "title": "SCAP policies",
          }
        }
      />
    </WithPermission>
  </Route>
  <Route
    exact={true}
    key="modal-route--scappolicies/:policy_id/edit"
    path="/scappolicies/:policy_id/edit"
  >
    <WithPermission
      requiredPermissions={
        Array [
          "compliance:*:*",
          "compliance:policy:update",
        ]
      }
    >
      <lazy
        route={
          Object {
            "component": Object {
              "$$typeof": Symbol(react.lazy),
              "_init": [Function],
              "_payload": Object {
                "_result": [Function],
                "_status": -1,
              },
            },
            "defaultTitle": "SCAP policies",
            "isCurrent": false,
            "modal": true,
            "path": "/scappolicies/:policy_id/edit",
            "requiredPermissions": Array [
              "compliance:*:*",
              "compliance:policy:update",
            ],
            "setTitle": [Function],
            "title": "$entityTitle - SCAP policies",
          }
        }
      />
    </WithPermission>
  </Route>
  <Route
    exact={true}
    key="modal-route--scappolicies/:policy_id/delete"
    path="/scappolicies/:policy_id/delete"
  >
    <WithPermission
      requiredPermissions={
        Array [
          "compliance:*:*",
          "compliance:policy:delete",
        ]
      }
    >
      <lazy
        route={
          Object {
            "component": Object {
              "$$typeof": Symbol(react.lazy),
              "_init": [Function],
              "_payload": Object {
                "_result": [Function],
                "_status": -1,
              },
            },
            "isCurrent": false,
            "modal": true,
            "path": "/scappolicies/:policy_id/delete",
            "requiredPermissions": Array [
              "compliance:*:*",
              "compliance:policy:delete",
            ],
            "setTitle": [Function],
            "title": "Delete policy - SCAP policies",
          }
        }
      />
    </WithPermission>
  </Route>
  <Route
    exact={true}
    key="modal-route--reports/:report_id/delete"
    path="/reports/:report_id/delete"
  >
    <WithPermission
      requiredPermissions={
        Array [
          "compliance:*:*",
          "compliance:report:delete",
        ]
      }
    >
      <lazy
        route={
          Object {
            "component": Object {
              "$$typeof": Symbol(react.lazy),
              "_init": [Function],
              "_payload": Object {
                "_result": [Function],
                "_status": -1,
              },
            },
            "isCurrent": false,
            "modal": true,
            "path": "/reports/:report_id/delete",
            "requiredPermissions": Array [
              "compliance:*:*",
              "compliance:report:delete",
            ],
            "setTitle": [Function],
            "title": "Delete report - Reports",
          }
        }
      />
    </WithPermission>
  </Route>
  <Route
    exact={true}
    key="modal-route--reports/:report_id/pdf"
    path="/reports/:report_id/pdf"
  >
    <WithPermission
      requiredPermissions={
        Array [
          "compliance:*:*",
          "compliance:report_export:read",
        ]
      }
    >
      <lazy
        route={
          Object {
            "component": Object {
              "$$typeof": Symbol(react.lazy),
              "_init": [Function],
              "_payload": Object {
                "_result": [Function],
                "_status": -1,
              },
            },
            "defaultTitle": "Reports",
            "isCurrent": false,
            "modal": true,
            "path": "/reports/:report_id/pdf",
            "requiredPermissions": Array [
              "compliance:*:*",
              "compliance:report_export:read",
            ],
            "setTitle": [Function],
            "title": "Export report - Reports",
          }
        }
      />
    </WithPermission>
  </Route>
</Suspense>
`;
